libsolidity/analysis/IntegerOverflowBounds.cpp

```diff
@@ -0,0 +1,267 @@
+	if (op == Token::Add)
+	{
+		min = left.min + right.min;
+		max = left.max + right.max;
+	}
+	else if (op == Token::Sub)
+	{
+		min = left.min - right.max;
+		max = left.max - right.min;
+	}
+	else if (op == Token::Mul)
+	{
+		dev::bigint products[4];
+		products[0] = left.min * right.min;
+		products[1] = left.min * right.max;
+		products[2] = left.max * right.min;
+		products[3] = left.max * right.max;
+
+		min = products[0];
+		max = products[0];
+
+		for (int i=1; i<4; i++)
+		{
+			if (products[i] < min)
+				min = products[i];
+			if (products[i] > max)
+				max = products[i];
+		}
+	}
+	else if (op == Token::Div)
+	{
+		// TODO: Add check for divide by zero
+		// we don't send the ErrorReporter to this level
+		// so not sure yet how to report it
+		//
+		// Formula is:
+		// [x1,x2]/[y1,y2] = [x1,x2]*(1/[y1,y2])
+		// Where, if [y1,y2] doesn't contain 0,
+		// 1/[y1,y2] = [1/y2,1/y1]
+		//
+		// Thus, we get [x1,x2]*[1/y2,1/y1].
+		// And since a*(1/b) = a/b, this should be
+		// the multiplication op, but with division
+		// and a slight reordering.
+		//
+		dev::bigint quotients[4];
+		quotients[0] = left.min / right.max;
+		quotients[1] = left.min / right.min;
+		quotients[2] = left.max / right.max;
+		quotients[3] = left.max / right.min;
+
+		min = quotients[0];
+		max = quotients[0];
+
+		for (int i = 1; i < 4; i++)
+		{
+			if (quotients[i] < min)
+				min = quotients[i];
+			if (quotients[i] > max)
+				max = quotients[i];
+		}
+	}
+	else if (op == Token::Mod)
+	{
+		min = left.min % right.min;
+		max = left.max % right.max;
+		swapIfNecessary();
+	}
+	// The < and > cases are most assuredly wrong in the general case.
+	// They fail to cover scenarios like [5,6] < [2,3], which is quite
+	// tautologically false.  Scenarios like that should actually generate
+	// their own warning or something and they could seriously impact
+	// how we can create and apply constraints.
+	//
+	// For now, the LHS passed in here is always the widest range for the type
+	// so this is safe-ish but still wrong.
+	else if (op == Token::LessThan)
+	{
+		max = right.min-1;
+	}
+	else if (op == Token::GreaterThan)
+	{
+		min = right.max+1;
+	}
+	else if (op == Token::Equal)
+	{
+		min = right.min;
+		max = right.max;
+	}
+}
+
+void IntegerOverflowBounds::swapIfNecessary()
+{
+	if (min > max)
+	{
+		dev::bigint swap = min;
+		min = max;
+		max = swap;
+	}
+}
+
+bool IntegerOverflowBounds::fitsType(Type const& type)
+{
+	if (type.category() == Type::Category::Integer)
+		return fitsType(static_cast<IntegerType const&>(type));
+	solAssert(false, "IntegerOverflowBounds fits only IntegerType");
+}
+
+bool IntegerOverflowBounds::fitsType(IntegerType const& type)
+{
+	bool isSigned = type.isSigned();
+	uint16_t bytes = type.storageBytes();
+	uint32_t exponent = bytes * (isSigned ? ((bytes * 8) - 1) : (bytes * 8));
+	dev::bigint allowedMin;
+	dev::bigint allowedMax;
+	if (isSigned)
+	{
+		allowedMin = 1;
+		allowedMin = allowedMin << exponent;
```

Comment:
I'm not against it.  I want to do a code reorg before I proceed.  When you say "each type," do you mean IntegerType and RationalNumberType?